util/topgen.py

```diff
@@ -287,6 +287,42 @@ def generate_plic(top, out_path):
     with rtl_gen_path.open(mode='w', encoding='UTF-8') as fout:
         fout.write(genhdr + gencmd + out)
 
+def _calc_pin_pos(top, mname):
+    dios = top["pinmux"]["dio"]
+
+    last_index = dios.index(dios[-1])
+    bit_pos = []
+    first_index = False
+
+    for dio in dios:
+        if dio['module_name'] == mname and not first_index:
+            bit_pos.append(last_index - dios.index(dio))
+            first_index = True
+        elif first_index and dio['module_name'] != mname:
+            bit_pos.append(last_index - dios.index(dio))
+
+    # The last one, need to insert last element
+    if len(bit_pos) == 1:
+        bit_pos.append(0)
```

Comment:
I am worrying about off-by-one issues here but not sure your intention for bit pos. Does the final index represent the last signal that is in mname or the first one that is not? There may need to be a +1 in your elif?

Suppose the last two entries are from mname then we get bit_pos = 1:0 with the 0 coming from the fall off the end check, but if the penultimate two are from mname and the final is not then the result will be 2:0 because the elif dios.index(dio) will be last_index.

Reply:
Looking at the generated top, I think the USB are the last dios which is the case that the code will get correct.

Reply:
yeah good point, i think the `elif` case does have an off by one issue.
Let me fix that.  I think in this case for usb since it happened to be the last two, it only hit the `if` and the `last one` scenarios.